pkg/nsx/services/expand.go

```diff
@@ -0,0 +1,277 @@
+		log.V(2).Info("built ruleIPSetGroup", "ruleIPSetGroup", ruleIPSetGroup)
+		nsxGroups = append(nsxGroups, ruleIPSetGroup)
+	}
+	log.V(2).Info("built rule by service entry", "rule", nsxRule)
+	return nsxGroups, nsxRule, nil
+}
+
+// Resolve a named port to port number by rule and policy selector.
+// e.g. "http" -> [{"80":['1.1.1.1', '2.2.2.2']}, {"443":['3.3.3.3']}]
+func (service *SecurityPolicyService) resolveNamedPort(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule,
+	spPort v1alpha1.SecurityPolicyPort) ([]util.Address, error) {
+
+	var address []util.Address
+
+	podSelector, namespaces, err := service.getPodSelector(obj, rule)
+	if err != nil {
+		return nil, err
+	}
+	if len(namespaces) == 0 {
+		namespaces = append(namespaces, obj.Namespace)
+	}
+
+	for _, namespace := range namespaces {
+		podsList := &v1.PodList{}
+		podSelector.Namespace = namespace
+		log.V(2).Info("port", "podSelector", podSelector)
+		err := service.Client.List(context.Background(), podsList, podSelector)
+		if err != nil {
+			return nil, err
+		}
+		for _, pod := range podsList.Items {
+			addr, err := service.resolvePodPort(&pod, &spPort)
+			switch err.(type) {
+			default:
+			case util2.PodIPNotFound, util2.PodNotRunning:
+				return nil, err
+			}
+			address = append(address, addr...)
+		}
+	}
+
+	if len(address) == 0 {
+		return nil, util2.NoFilteredPod{
+			Desc: "no pod has the corresponding named port, perhaps pod has not been created yet",
+		}
+	}
+	return util.MergeAddressByPort(address), nil
+}
+
+// Check port name and protocol, only when the pod is really running, and it does have effective ip.
+func (service *SecurityPolicyService) resolvePodPort(pod *v1.Pod, spPort *v1alpha1.SecurityPolicyPort) (
+	[]util.Address, error) {
+	var addr []util.Address
+	for _, container := range pod.Spec.Containers {
+		for _, port := range container.Ports {
+			log.V(2).Info("resolvePodPort", "namespace", pod.Namespace, "pod_name", pod.Name,
+				"port_name", port.Name, "containerPort", port.ContainerPort,
+				"protocol", port.Protocol, "podIP", pod.Status.PodIP)
+			if port.Name == spPort.Port.String() && port.Protocol == spPort.Protocol {
+				if pod.Status.Phase != "Running" {
+					errMsg := fmt.Sprintf("pod %s/%s is not running", pod.Namespace, pod.Name)
+					return nil, util2.PodNotRunning{Desc: errMsg}
+				}
+				if pod.Status.PodIP == "" {
+					errMsg := fmt.Sprintf("pod %s/%s ip not initialized", pod.Namespace, pod.Name)
+					return nil, util2.PodIPNotFound{Desc: errMsg}
+				}
+				addr = append(addr, util.Address{Port: int(port.ContainerPort), IPs: []string{pod.Status.PodIP}})
+			}
+		}
+	}
+	return addr, nil
+}
+
+// Build an ip set group for NSX.
+func (service *SecurityPolicyService) buildRuleIPGroup(obj *model.Rule, ips []string) *model.Group {
+	ipGroup := model.Group{}
+
+	policyGroupID := fmt.Sprintf("%s_ipset", *obj.Id)
+	ipGroup.Id = &policyGroupID
+	policyGroupName := fmt.Sprintf("%s-ipset", *obj.DisplayName)
```

Comment:
same here, is policyGroupName shold be IPSetGroupName?

Reply:
good